### PR TITLE
Support to resize uploaded volume during attach

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR addresses #5906  

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?
Uploaded a volume with custom disk offering, 
Attach the volume to a VM, it performs a resize operation on attach.

### CODE REVIEW

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public VolumeInfo getVolume() {
 
     AsyncCallFuture<VolumeApiResult> resize(VolumeInfo volume);
 
-    void resizeVolumeOnHypervisor(long volumeId, long newSize, long destHostId, String instanceName);
+    void resizeVolumeOnHypervisor(long volumeId, long currentSize, long newSize, long destHostId, String instanceName);
```
Comment (DaanHoogland) [inline engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: why is the parameter be called `currentSize` needed? seems like `newSize` is enough.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -353,9 +354,23 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+            if (sizeInGB == null) {
+                throw new InvalidParameterValueException("This disk offering requires a custom size specified");
+            }
+            Long customDiskOfferingMaxSize = VolumeOrchestrationService.CustomDiskOfferingMaxSize.value();
+            Long customDiskOfferingMinSize = VolumeOrchestrationService.CustomDiskOfferingMinSize.value();
+
+            if ((sizeInGB < customDiskOfferingMinSize) || (sizeInGB > customDiskOfferingMaxSize)) {
+                throw new InvalidParameterValueException("Volume size: " + sizeInGB + "GB is out of allowed range. Max: " + customDiskOfferingMaxSize + " Min:" + customDiskOfferingMinSize);
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:373]: can you extract this and give a good merthod name?

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 3]
```diff
@@ -364,7 +379,27 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+            volume = _volsDao.findById(vol.getId());
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
+                    _volsDao.persist(volume);
+                }
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException(String.format("Failed to register volume due to - %s", e.getMessage()), e);
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:402]: can you extract?

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java (hunk 1 above)
Comment (Pearl1594) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: Yes it should suffice as currentsize can be got from volume. But in case of an uploaded volume with a custom disk offering specified, we set the new size (i.e., the size that it would get resized to eventually during attach) as the volume size, instead of the physical size. And since I am reusing this method to resize the volume after the attach operation, vol.size() = currentSize would be equal to newSize.

Comment (Pearl1594) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: I added the parameter because, for upload volume, currentSize would be equal to newSize (as both would point to volume.getSize()) if currentSize wasn't explicitly passed as the size of the volume on the secondary store.

Comment (DaanHoogland) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: I still don't get why it has to be passed.

Comment (Pearl1594) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: So currently, https://github.com/apache/cloudstack/blob/9eefc76fc5ef9abf3d74a5cc6c528ca404c707fd/engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java#L2429 - uses volume.getSize() as the current Size and newSize is what ever size we want the volume to get resized to. In case of resizing an attached volume, volume.getSize() would be the size to which volume would be resized to. And if I weren't to explicitly pass the actual size of the volume (as it was when on the secondary store) currentSize would be equal to newSize and hence resizing wouldn't happen. Because, with this PR I update the size of the uploaded volume to the size that the disk offering enforces https://github.com/apache/cloudstack/pull/6005/files#diff-b93da5c5293f8cb4470ff6b1642d312131bb053df07b71b4b951b08dff0ad9abR404-R405

Comment (DaanHoogland) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java:95]: ok, thanks for explaining. I think this is a problem that needs to be addressed for all hypervisors separately. The old size should be retrieved from the hypervisor/primary storage. What we pass here could be off for any number of reasons. This is a bit out of scope here, but can you add a comment to what the problem is?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Hi @Pearl1594 it seems the failing test on test_volumes needs fixing, can you please check? The other failing test seems unrelated

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutanpackage

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan help

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 4]
```diff
@@ -364,7 +383,31 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+                s_logger.warn("Failed to register the volume");
+                String details = "";
+                if (result.getResult() != null && !result.getResult().isEmpty()) {
+                    details = result.getResult();
+                }
+                throw new CloudRuntimeException(details);
+            }
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:404]: At this point is there any other reference that needs to be updated, like volume_store_ref? Could this cause any issues with GetVolumeStats or not considered at this volume state?

**Comment by nvazquez** [comment]
XenServer failures still needs to be addressed, moving to 4.18

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 5]
```diff
@@ -2269,8 +2316,19 @@ public Volume attachVolumeToVM(Long vmId, Long volumeId, Long deviceId) {
...
         }
+
+        Long hostId = null;
+        if (vm.getHostId() != null) {
+            hostId = vm.getHostId();
+        } else if (vm.getLastHostId() != null) {
+            hostId = vm.getLastHostId();
+        }
+
+        if (vol != null && storeSize != null && hostId != null) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2328]: We can use `ObjectUtils#allNotNull` here.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 6]
```diff
@@ -364,7 +383,31 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+                throw new CloudRuntimeException(details);
+            }
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
+                    _volsDao.persist(volume);
+                }
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException(String.format("Failed to register volume due to - %s", e.getMessage()), e);
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:409]: We can add some info about the volume here.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 7]
```diff
@@ -337,6 +338,20 @@ protected VolumeApiServiceImpl() {
         _gson = GsonHelper.getGsonLogger();
     }
 
+    private void validateDiskSize(DiskOfferingVO diskOffering, Long sizeInGB) {
+        if (diskOffering != null && diskOffering.isCustomized()) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:342]: We can invert the if and add a return statement to reduce code indentation.

**Comment by DaanHoogland** [comment]
@Pearl1594 you have a conflict :(

**Comment by rohityadavcloud** [comment]
@Pearl1594 can you resolve the conflicts?

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 8]
```diff
@@ -377,7 +396,31 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+                    details = result.getResult();
+                }
+                throw new CloudRuntimeException(details);
+            }
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
+                    _volsDao.persist(volume);
+                }
+            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:420]:
```suggestion
            if (! (sizeInGB ==null || Objects.equals(sizeInGB, vol.getSize())) {
                volume.setSize(sizeInGB * GiB_TO_BYTES);
                _volsDao.persist(volume);
            }
```

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 6 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:422]: @GutoVeronezi what do you expect? Keeping in mind that the upload/registering failed, so effectively there is no volume to speak about.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 9]
```diff
@@ -377,7 +396,29 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
...
+            VolumeApiResult result = future.get();
+            if (result.isFailed()) {
+                s_logger.warn("Failed to register the volume");
+                String details = "";
+                if (result.getResult() != null && !result.getResult().isEmpty()) {
+                    details = result.getResult();
+                }
+                throw new CloudRuntimeException(details);
+            }
+            if (! (sizeInGB ==null || Objects.equals(sizeInGB, vol.getSize())) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:415]:
```suggestion
            if (! (sizeInGB ==null || Objects.equals(sizeInGB, vol.getSize()))) {
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepenv

**Comment by DaanHoogland** [comment]
@blueorangutan test keepenv

**Comment by weizhouapache** [comment]
> Please note that this change is only on the API level - no UI. There is a new parameter 'size' to the 'upload volume' command and you're able to customize the volume size during upload. The problem is I was able to upload a 1.3 GB image in a 1 GB data volume - I think there must be a check for this. Can be reproduced with the following cloudmonkey command:
> 
> upload volume name=centos7 url=https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1706.qcow2 zoneid= diskofferingid= format=QCOW2 size=1

it might be different to check the size before the template is uploaded (in ssvm).
the workaround might be: after the template upload, if the new size is smaller than actual size, use the actual size as volume size, otherwise use the new size.
cc @DaanHoogland @vladimirpetrov

**Comment by DaanHoogland** [comment]
> > Please note that this change is only on the API level - no UI. There is a new parameter 'size' to the 'upload volume' command and you're able to customize the volume size during upload. The problem is I was able to upload a 1.3 GB image in a 1 GB data volume - I think there must be a check for this. Can be reproduced with the following cloudmonkey command:
> > upload volume name=centos7 url=https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1706.qcow2 zoneid= diskofferingid= format=QCOW2 size=1
> 
> it might be different to check the size before the template is uploaded (in ssvm). the workaround might be: after the template upload, if the new size is smaller than actual size, use the actual size as volume size, otherwise use the new size. cc @DaanHoogland @vladimirpetrov

agreed

**Comment by weizhouapache** [comment]
@DaanHoogland 
we need to revisit this PR after 4.18 release.

I have added the UI support and fixed a critical bug.

However , it still does not work. the issues are
(1) in DB/UI, the volume has new size (I set to 11GB), but on secondary storage, it is still the old size (8GB in my testing)
```
# qemu-img info /mnt/secstoragetmp/volumes/2/498/b10834c2-8b11-33a5-add9-a2cc68f7f148.qcow2 
image: /mnt/secstoragetmp/volumes/2/498/b10834c2-8b11-33a5-add9-a2cc68f7f148.qcow2
file format: qcow2
virtual size: 8 GiB (8589934592 bytes)
disk size: 1.65 GiB
cluster_size: 65536
Format specific information:
    compat: 0.10
    compression type: zlib
    refcount bits: 16
```

(2) after attaching it to a vm, the volume is copied to primary storage. however, the size is still not changed (=old size)

```
# qemu-img info /mnt/657524e2-065b-3ca8-9e97-746b3c7d17db/2792282d-1e69-40a7-9072-206c7ff133f5
image: /mnt/657524e2-065b-3ca8-9e97-746b3c7d17db/2792282d-1e69-40a7-9072-206c7ff133f5
file format: qcow2
virtual size: 8 GiB (8589934592 bytes)
disk size: 1.65 GiB
cluster_size: 65536
Format specific information:
    compat: 0.10
    compression type: zlib
    refcount bits: 16
```

In short, volume is never resized (on KVM). I did not see any log to resize the volume.

**Comment by DaanHoogland** [comment]
@weizhouapache any way forward with this? Should we visit this on thursday?

**Comment by rohityadavcloud** [comment]
ping @Pearl1594 is this still relevant?

**Comment by Pearl1594** [comment]
I believe it is still relevant, it was more of an enhancement request. I think there is an issue based on Wei's comment, I'll need to spend some time on it.

**Comment by weizhouapache** [comment]
moved to 4.19.0.0 milestone as this is an enhancement

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
Is this worth rebooting @Pearl1594 and can you advise for 4.18 or 4.19/main?

**Comment by rohityadavcloud** [comment]
@Pearl1594 any update on this?

**Comment by DaanHoogland** [comment]
ping @Pearl1594

**Comment by DaanHoogland** [comment]
@Pearl1594 @sureshanaparti , closing this one as it has conflicts and is old. please update and reopen if you think it is still relevant.
